src/vsr/README.md

```diff
@@ -0,0 +1,184 @@
+
+    R0->>+R0: Prepare A
+    R0->>+R1: Prepare A
+    R1->>+R2: Prepare A
+
+    R0->>-R0: Prepare-Ok A
+    R1->>-R0: Prepare-Ok A
+    R0->>C0: Reply A
+    R2->>-R0: Prepare-Ok A
+```
+
+See also:
+
+  - [VRR](https://pmg.csail.mit.edu/papers/vr-revisited.pdf) §4.1
+
+## Protocol: Start-View-Change
+
+Start-View-Change (SVC) protocol initiates [view-changes](#protocol-view-change) with minimal disruption.
+
+Unlike the Start-View-Change described in [VRR](https://pmg.csail.mit.edu/papers/vr-revisited.pdf) §4.2, this protocol runs in both `status=normal` and `status=view_change` (not just `status=view_change`).
+
+1. Depending on the replica's status:
+    - `status=normal` & primary: When the replica has not recently received a `prepare_ok` (and it has a prepare in flight), pause broadcasting `command=commit`.
+    - `status=normal` & backup: When the replica has not recently received a `command=commit`, broadcast `command=start_view_change` to all replicas (including self).
+    - `status=view_change`: If the replica has not completed a view-change recently, send a `command=start_view_change` to all replicas (including self).
+2. (Periodically retry sending the SVC).
+3. If the backup receives a `command=commit` or changes views (respectively), stop the `command=start_view_change` retries.
+4. If the replica collects a [view-change quorum](#quorums) of SVC messages, transition to `status=view_change` for the next view. (That is, increment the replica's view and start sending a DVC).
+
+This protocol approach enables liveness under asymmetric network partitions. For example, a replica which can send to the cluster but not receive may send SVCs, but if the remainder of the cluster is healthy, they will never achieve a quorum, so the view is stable. When the partition heals, the formerly-isolated replica may rejoin the original view (if it was isolated in `status=normal`) or a new view (if it was isolated in `status=view_change`).
+
+See also:
+
+  - [Raft does not Guarantee Liveness in the face of Network Faults](https://decentralizedthoughts.github.io/2020-12-12-raft-liveness-full-omission/) ("PreVote and CheckQuorum")
+  - ["Consensus: Bridging Theory and Practice"](https://web.stanford.edu/~ouster/cgi-bin/papers/OngaroPhD.pdf) §6.2 "Leaders" describes periodically committing a heartbeat to detect stale leaders.
+
+## Protocol: View-Change
+
+A replica sends `command=do_view_change` to all replicas, with the `view` it is attempting to start.
+- The _primary_ of the `view` collects a [view-change quorum](#quorums) of DVCs.
+- The _backup_ of the `view` uses to `do_view_change` to updates its current `view` (transitioning to `status=view_change`).
+
+When the primary collects its quorum:
+1. The primary installs the headers to its suffix.
+2. Then the primary repairs (and potentially truncates uncommitted ops).
+3. When the primary finishes repair, it transitions to `status=normal` and broadcasts a `command=start_view`.
```

Comment:
Could you explain in a bit more detail how this happens, and especially how this interacts with repairs?  This is super-fuzzy for me at the moment. Some aspects I understand are:

* there are headers (small, needed to establish what's committed) and preparess (large, needed to actually apply state transitions). 
* we assume they might get corrupted
* DVC includes only headers, although they might bee corrupted (?)
* Upon receiving DVCs, primary needs to reconstruct what's actually commited using view-stamps
* It also needs to do something about "corrupted" entries (if we have a quorum of DVC, but it has some corrupted entries, we don't know if they are committed or not)
* there's nacking process, which allows the primary to figure out if something is definitely _not_ committed (why do we need nacks here? Can't we figure this out from DVCs alone?)

Reply:
> (why do we need nacks here? Can't we figure this out from DVCs alone?)

answering my question: DVC gives us a quorum, which the primary is a part of. But if a record on the primary is corrupted, there might not be a quorum for this record! So the primary would have to ask someone outside of DVC about this record.

In other words, quorums guarantee us that “there exists a node with the record”. But if the record is corrupted on that node, we need to look beyound the quorum!

Reply:
There are 3 possible states for the record (in the WAL), for each replica:

1. Present + valid
2. Corrupt
3. Missing

We need to distinguish those to determine which ops are uncommitted, and can/must be truncated to ensure availability. In all 3 of those cases it is possible that the header is in the replica's journal, and therefore the header is included in the DVC. The nack step's job is to distinguish these cases, and identify which can be truncated. Specifically, in response to a `request_prepare` (during a view change):

1. Present + valid → replica replies with the prepare itself.
2. Corrupt → replica does not return a reply at all
3. Missing → replica replies with nack

Consider this example:

1. `replica_count=3`
2. Replica 0 is leader and prepares 4 ops in parallel (pipelining): `op=A, op=B, op=C, op=D`.
3. Replica 0 replicates ops `A,B,C,D` to replica 1.
4. Replica 1 receives + prepares ops `B,C,D`. (Also forwards to replica 2).
5. Replica 1's `repair_timeout` requests the headers to fix the gap, so now it has header `A`.
6. Replica 2 receives + prepares ops `B,C,D`.
7. Replica 0 crashes.
8. Replica 1 is the new leader:
    - It has the full set of headers: `A,B,C,D`.
    - But it never received `A`.
    - It asks replica 2 for `A`, but replica 2 didn't receive `A` either (`command=nack_prepare`).
    - Since 2/3 _never prepared_/_acked_ `A`, it knows `A` is uncommitted.
    - Therefore `B,C,D` are uncommitted.
    - So it can truncate `B,C,D` and transition to `status=normal`.


> DVC includes only headers, although they might bee corrupted (?)

Just to be clear, the headers themselves are not corrupted (if they were, they would not be included in the DVC). But the prepares (on-disk, in the WAL) corresponding to the headers may be corrupt.